docs: `ScheduleCreateTransaction` (#524)
hiero-ledger/hiero-sdk-tck: docs/test-specifications/schedule-service/ScheduleCreateTransaction.md

| memo                    | string                                                  | optional          | Short description of the schedule (UTF-8 encoding max 100 bytes)                                                                            |
| adminKey                | string                                                  | optional          | DER-encoded hex string representation for private or public keys. Keylists and threshold keys are the hex of the serialized protobuf bytes. |
| payerAccountId          | string                                                  | optional          | ID of the account to pay for the scheduled transaction.                                                                                     |
| expirationTime          | string                                                  | optional          | The time at which this schedule should expire (in seconds since the epoch)                                                                  |

> | waitForExpirty          | bool                                                    | optional          | Should the schedule wait until its expiration time to execute?                                                                              |

Reviewer: waitForExpiry

Author: fixed

| scheduleId     | string | The ID of the created schedule.                                                        |
| status         | string | The status of the submitted `ScheduleCreateTransaction` (from a `TransactionReceipt`). |

### Additional Notes

The tests contained in this specification will assume that scheduled transaction is a basic account create transaction with an ED25519 public key and it will be represented by a <SCHEDULED_TRANSACTION> tag, unless the test specifies otherwise.

## Property Tests

### **Scheduled Transaction:**

- The transaction that is to be scheduled.

| Test no | Name                                                                 | Input                                              | Expected response                                                                                           | Implemented (Y/N) |
|---------|----------------------------------------------------------------------|----------------------------------------------------|-------------------------------------------------------------------------------------------------------------|-------------------|

> | 1       | Creates a scheduled transaction                                      | scheduledTransaction=<SCHEDULED_TRANSACTION>       | The schedule creation succeeds and the transaction is scheduled.                                            | N                 |

Author: Added

| 2       | Creates a scheduled transaction with no scheduled transaction        |                                                    | The schedule creation fails with an `INVALID_TRANSACTION_BODY` response code from the network.              | N                 |

> | 3       | Creates a scheduled transaction that's not a whitelisted transaction | scheduledTransaction=<NON_WHITELISTED_TRANSACTION> | The schedule creation fails with a `SCHEDULED_TRANSACTION_NOT_IN_WHITELIST` response code from the network. | N                 |

Author: I'm not sure if that's something we'd want the SDK to do.. if more transactions are whitelisted later, users would need to update their SDKs to the latest version to be able to schedule those transactions

> | 3       | Creates a scheduled transaction that's not a whitelisted transaction | scheduledTransaction=<NON_WHITELISTED_TRANSACTION> | The schedule creation fails with a `SCHEDULED_TRANSACTION_NOT_IN_WHITELIST` response code from the network. | N                 |

Author: For the transactions that don't have a scheduable body, we should just throw. But I think it makes sense for the broader set to allow the network to decide what transactions can be scheduled and what can't

> | 3       | Creates a scheduled transaction that's not a whitelisted transaction | scheduledTransaction=<NON_WHITELISTED_TRANSACTION> | The schedule creation fails with a `SCHEDULED_TRANSACTION_NOT_IN_WHITELIST` response code from the network. | N                 |

Reviewer 2: Yeah i think that's the case for the rest.

### **Wait for Expiry:**

- Should the scheduled transaction wait until the expiration time to execute?

| Test no | Name                                                | Input                                                             | Expected response                                                                             | Implemented (Y/N) |
|---------|-----------------------------------------------------|-------------------------------------------------------------------|-----------------------------------------------------------------------------------------------|-------------------|
| 1       | Creates a schedule with wait for expiry             | scheduledTransaction=<SCHEDULED_TRANSACTION>, waitForExpiry=true  | The schedule creation succeeds and the transaction will wait until expiration to execute.     | N                 |
| 2       | Creates a schedule with expiration time in the past | scheduledTransaction=<SCHEDULED_TRANSACTION>, waitForExpiry=false | The schedule creation succeeds and the transaction will not wait until expiration to execute. | N                 |

> 

Reviewer 2: I think we would hit the max transaction limit and this would cause problems because the grpc compilers in different languages can create different size messages for - on a first glance - the same transaction. I think it would be tricky to pull off.

| expirationTime          | string                                                  | optional          | The time at which this schedule should expire (in seconds since the epoch)                                                                  |
| waitForExpiry           | bool                                                    | optional          | Should the schedule wait until its expiration time to execute?                                                                              |
| commonTransactionParams | [json object](../common/CommonTransactionParameters.md) | optional          |                                                                                                                                             |

### Output Parameters

| Parameter Name | Type   | Description/Notes                                                                      |
|----------------|--------|----------------------------------------------------------------------------------------|

> | scheduleId     | string | The ID of the created schedule.                                                        |

Author: added

| 2       | Creates a scheduled consensus submit message transaction             | scheduledTransaction=<CONSENSUS_SUBMIT_MESSAGE_TRANSACTION> | The schedule creation succeeds and the transaction is scheduled.                                            | N                 |
| 3       | Creates a scheduled token burn transaction                           | scheduledTransaction=<TOKEN_BURN_TRANSACTION>               | The schedule creation succeeds and the transaction is scheduled.                                            | N                 |
| 4       | Creates a scheduled token mint transaction                           | scheduledTransaction=<TOKEN_MINT_TRANSACTION>               | The schedule creation succeeds and the transaction is scheduled.                                            | N                 |
| 5       | Creates a scheduled crypto approve allowance transaction             | scheduledTransaction=<CRYPTO_APPROVE_ALLOWANCE_TRANSACTION> | The schedule creation succeeds and the transaction is scheduled.                                            | N                 |

> | 6       | Creates a scheduled transaction with no scheduled transaction        |                                                             | The schedule creation fails with an `INVALID_TRANSACTION` response code from the network.                   | N                 |

Reviewer 2: This test can be split into 2 - one using 0.0.2 and one using normal account. When using 0.0.2 it returns `INVALID_TRANSACTION` but if not - it fails with FAIL_INVALID which need to be communicated with the team.

Author: For the payer account? Or as the operator?

Author: Confirmed with the services team that `FAIL_INVALID` is a bug and that the correct response code should always be `INVALID_TRANSACTION`, so I don't think we need two tests here.

### **Payer Account ID:**

- The ID of the account to pay for the scheduled transaction.

| Test no | Name                                                          | Input                                                                                                                                              | Expected response                                                                                      | Implemented (Y/N) |
|---------|---------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------|-------------------|
| 1       | Creates a schedule with a payer account ID                    | scheduledTransaction=<SCHEDULED_TRANSACTION>, payerAccountId=<CREATED_ACCOUNT_ID>, commonTransactionParams.signers=[<ACCOUNT_PRIVATE_KEY>]         | The schedule creation succeeds and the transaction has the account ID as its payer account ID.         | N                 |

> | 2       | Creates a schedule with a payer account ID that doesn't exist | scheduledTransaction=<SCHEDULED_TRANSACTION>, payerAccountId="123.456.789"                                                                         | The schedule creation succeeds and the transaction has the invalid account ID as its payer account ID. | N                 |

Reviewer 2: ```suggestion
| 2       | Creates a schedule with a payer account ID that doesn't exist | scheduledTransaction=<SCHEDULED_TRANSACTION>, payerAccountId="123.456.789"                                                                         | Transaction fails with `ACCOUNT_ID_DOES_NOT_EXIST`. | N                 |
```

This actually fails

| 7       | Creates a schedule with expiration time of -9,223,372,036,854,775,808 (`int64` min) seconds     | scheduledTransaction=<SCHEDULED_TRANSACTION>, expirationTime="-9223372036854775808"   | The schedule creation fails with `SCHEDULE_EXPIRATION_TIME_MUST_BE_HIGHER_THAN_CONSENSUS_TIME`. | N                 |
| 8       | Creates a schedule with expiration time of -9,223,372,036,854,775,807 (`int64` min + 1) seconds | scheduledTransaction=<SCHEDULED_TRANSACTION>, expirationTime="-9223372036854775807"   | The schedule creation fails with `SCHEDULE_EXPIRATION_TIME_MUST_BE_HIGHER_THAN_CONSENSUS_TIME`. | N                 |
| 9       | Creates a schedule with expiration time of 5,356,800 seconds from the current time              | scheduledTransaction=<SCHEDULED_TRANSACTION>, expirationTime=<CURRENT_TIME + 5356800> | The schedule creation succeeds and the file has the specified expiration time.                  | N                 |

> | 10      | Creates a schedule with expiration time of 5,356,801 seconds from the current time              | scheduledTransaction=<SCHEDULED_TRANSACTION>, expirationTime=<CURRENT_TIME + 5356801> | The schedule creation fails with `SCHEDULE_EXPIRATION_TIME_TOO_FAR_IN_FUTURE`.                  | N                 |

Reviewer 2: ```suggestion
| 10      | Creates a schedule with expiration time of 5,356,801 seconds from the current time              | scheduledTransaction=<SCHEDULED_TRANSACTION>, expirationTime=<CURRENT_TIME + 5356801> | The schedule creation fails with `SCHEDULE_EXPIRATION_TIME_TOO_FAR_IN_FUTURE`.                  | N                 |
```

I am concerned if this will be flaky since the time.now when running the test will be different from time.now in the consensus node itself.

Author: Gave a couple seconds buffer
